iwyu.cc

```diff
@@ -1375,62 +1375,70 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
+    SourceLocation spelling_loc = sm->getSpellingLoc(loc);
+
+    VERRS(5) << "Original use loc: " << PrintableLoc(use_loc) << "\n";
+    VERRS(5) << "Spelling loc: " << PrintableLoc(spelling_loc) << "\n";
+    VERRS(5) << "Expansion loc: " << PrintableLoc(expansion_loc) << "\n";
+
+    const FileEntry* macro_def_file = GetLocFileEntry(spelling_loc);
+    VERRS(5) << "Macro is defined in file '" << GetFilePath(macro_def_file)
+             << "'. Looking for fwd-decl hint...\n";
+
     const NamedDecl* fwd_decl = nullptr;
     for (const NamedDecl* redecl : GetClassRedecls(decl)) {
-      if (GetFileEntry(redecl) == macro_def_file &&
-          IsForwardDecl(redecl)) {
+      if (GetFileEntry(redecl) == macro_def_file && IsForwardDecl(redecl)) {
         fwd_decl = redecl;
+
+        // Make sure we keep that forward-declaration, even if it's probably
+        // unused in this file.
+        IwyuFileInfo* file_info =
+            preprocessor_info().FileInfoFor(macro_def_file);
+        file_info->ReportForwardDeclareUse(
+            spelling_loc, fwd_decl,
+            IsNodeInsideCXXMethodBody(current_ast_node()), nullptr);
         break;
       }
     }
 
-    if (fwd_decl) {
-      // Make sure we keep that forward-declaration, even if it's probably
-      // unused in this file.
-      preprocessor_info().FileInfoFor(macro_def_file)->ReportForwardDeclareUse(
-          use_loc, decl, IsNodeInsideCXXMethodBody(current_ast_node()),
-          nullptr);
-    }
-
-    // Resolve the best use location based on our current knowledge.
-    //
-    // 1) If the macro definition file forward-declares the used decl, that's a
-    //    hint that it wants the expansion location to take responsibility.
-    // 2) If the symbol being used is passed as an argument to the macro, we
-    //    want the expansion location to take responsibility (that thing they
-    //    passed in is something they know about, and the macro doesn't.)
-    // 3) If the use_loc is in <scratch space>, we assume it's formed by macro
-    //    argument concatenation, and attribute responsibility to the expansion
-    //    location.
-    //
-    // Otherwise, the macro file is responsible for including the symbol.
-    SourceLocation expansion_loc = GetInstantiationLoc(use_loc);
-    const char* responsible = "expansion";
-
     if (fwd_decl != nullptr) {
-      VERRS(5) << "Macro file has a forward-decl attribution hint.\n";
+      // If the macro file has a forward-decl, the expansion loc is responsible.
+      VERRS(5) << "Found a forward-decl in macro definition file.\n";
       use_loc = expansion_loc;
-    } else if (IsInScratchSpace(use_loc)) {
-      VERRS(5) << "Decl was used in <scratch space>, presumably as a result of "
-                  "macro arg concatenation.\n";
+    } else if (sm->isMacroArgExpansion(spelling_loc)) {
+      // For macro arguments, the expansion loc is responsible.
+      VERRS(5) << "Spelling location is in a macro-arg expansion.\n";
```

Comment:
Actually, the more precise test case to provoke this should be:

    // macro_location-d2.h
    #define CREATE_VAR(typ)   typ create_var

    // macro_location.h
    CREATE_VAR(HClass);

But I can see that the expansion is not working as I expected:

    tests/cxx/macro_location.h:26:1 <Spelling=./tests/cxx/macro_location-d2.h:37:31>: (1) [ VarDecl ] HClass create_var
    tests/cxx/macro_location.h:26:1 <Spelling=./tests/cxx/macro_location.h:26:12>: (2) [ RecordType ] class HClass     
    Trying to determine use location for 'class HClass {}'                                                             
    Original use loc: tests/cxx/macro_location.h:26:1 <Spelling=./tests/cxx/macro_location.h:26:12>                    
    Spelling loc: tests/cxx/macro_location.h:26:12                                                                     
    Expansion loc: tests/cxx/macro_location.h:26:12                                                                    
    Macro is defined in file 'tests/cxx/macro_location.h'. Looking for fwd-decl hint...                                
    Attributing use of 'class HClass {}' to location at tests/cxx/macro_location.h:26:12.                              

The original use location looks wrong here -- I would have expected the spelling loc to point into `macro_location-d2.h`. So maybe this is an issue with Clang's location handling, that implicitly locates macro arg expansions at some offset from the expansion loc?

I'll try and example with bad syntax inside a macro, and see if I can reproduce the error in Clang's error message.